src/Tutor.Core/UseCases/KnowledgeAnalysis/AssessmentAnalysisService.cs

```diff
@@ -0,0 +1,108 @@
+﻿using FluentResults;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Tutor.Core.BuildingBlocks;
+using Tutor.Core.BuildingBlocks.EventSourcing;
+using Tutor.Core.Domain.CourseIteration;
+using Tutor.Core.Domain.KnowledgeAnalysis;
+using Tutor.Core.Domain.KnowledgeMastery.Events;
+using Tutor.Core.Domain.KnowledgeMastery.Events.AssessmentItemEvents;
+using Tutor.Core.Domain.Stakeholders.RepositoryInterfaces;
+
+namespace Tutor.Core.UseCases.KnowledgeAnalysis;
+
+public class AssessmentAnalysisService : IAssessmentAnalysisService
+{
+    private readonly IGroupRepository _groupRepository;
+    private readonly IOwnedCourseRepository _ownedCourseRepository;
+    private readonly IEventStore _eventStore;
+
+    public AssessmentAnalysisService(IGroupRepository groupRepository, IOwnedCourseRepository ownedCourseRepository, IEventStore eventStore)
+    {
+        _groupRepository = groupRepository;
+        _ownedCourseRepository = ownedCourseRepository;
+        _eventStore = eventStore;
+    }
+
+    public Result<List<AiStatistics>> GetStatistics(int kcId, int instructorId)
+    {
+        if (!_ownedCourseRepository.IsKcOwner(kcId, instructorId)) return Result.Fail(FailureCode.Forbidden);
+
+        var events = _eventStore.Events
+            .Where(e => e.RootElement.GetProperty("KnowledgeComponentId").GetInt32() == kcId)
+            .ToList<KnowledgeComponentEvent>();
+
+        return CalculateStatistics(events);
+    }
+
+    public Result<List<AiStatistics>> GetStatisticsForGroup(int kcId, int groupId, int instructorId)
+    {
+        if (!_ownedCourseRepository.IsKcOwner(kcId, instructorId)) return Result.Fail(FailureCode.Forbidden);
+
+        var learnerIds = _groupRepository.GetLearnerIdsInGroup(groupId);
+        
+        var events = _eventStore.Events
+            .Where(e => e.RootElement.GetProperty("KnowledgeComponentId").GetInt32() == kcId)
+            .Where(e => learnerIds.Contains(e.RootElement.GetProperty("LearnerId").GetInt32()))
+            .ToList<KnowledgeComponentEvent>();
+
+        return CalculateStatistics(events);
+    }
+
+    private static List<AiStatistics> CalculateStatistics(List<KnowledgeComponentEvent> events)
+    {
+        var sortedAiEvents = events.OfType<AssessmentItemEvent>()
+            .OrderBy(e => e.TimeStamp).ToList();
```

Comment:
ToList calls here and onward seem not necessary, IEnumerable can be used as CreateAiStatistic param also.
Unnecessary ToList calls on lines:

- 56
- 67
- 86
- 96

Reply:
ReSharper warns me when multiple enumerations occur and so far I have adhered to this warning: https://www.jetbrains.com/help/resharper/PossibleMultipleEnumeration.html

I haven't really done a deep dive to understand the issue to its fullest, so it might be unneeded.